### PR TITLE
Use lockless mode when adding index on Azure Sql server

### LINKED ISSUE #6099: Migrations adding an index on Azure Sql Server lock the whole table
**Describe the bug**
Migrations lock the tables during upgrade blocking any writes to the system.

Adding an index to Sql Server without locking the table was traditionally and "Enterprise" feature.
Using the special syntax WITH (ONLINE = ON) is illegal on Developer edition.
We do NOT currently use this syntax on Azure cloud Sql Server despite it being supported.

**To Reproduce**

- Install Hapi JPA on 7.0 Azure Sql Server (docker is fine https://hub.docker.com/r/microsoft/azure-sql-edge)
- Then run a recent hapiutil migrate-database  --dry-run ...connection params..
- Inspect the migration output and look for the sql statements with create index
- Notice statements do not contain `WITH (ONLINE = ON)`

**Expected behavior**

- Sql statements should contain `WITH (ONLINE = ON)`

**Environment (please complete the following information):**
 - Hapi 7.3.8

### PR DESCRIPTION
**Pull request by michaelabuckley**
Fixed #6099

### CODE REVIEW

**Review by tadgh — APPROVED** [review]

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/MetadataSource.java [hunk 1]
```diff
@@ -37,11 +37,17 @@ public boolean isOnlineIndexSupported(DriverTypeEnum.ConnectionProperties theCon
 			case COCKROACHDB_21_1:
 				return true;
 			case MSSQL_2012:
+				// use a deny-list instead of allow list, so we have a better failure mode for new/unknown versions.
+				// Better to fail in dev than run with a table lock in production.
 				String mssqlEdition = getEdition(theConnectionProperties);
-				return mssqlEdition.startsWith("Enterprise");
+				return mssqlEdition == null // some weird version without an edition?
```
Comment (fil512) [inline hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/MetadataSource.java:43]: This doesn't look like the change you initially proposed
I thought your recommendation was to try offline mode, and if it failed then try again online.
Or maybe I'm not seeing the logic flow...?

Comment (michaelabuckley) [reply hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/MetadataSource.java:44]: I've replaced this with try-catch in the sql.
See SqlServerStandardMigrationTest, etc. for tests that verify this works.

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTask.java [hunk 2]
```diff
@@ -161,24 +162,62 @@ String generateSql() {
 					setTransactional(false);
 					break;
 				case ORACLE_12C:
+					// todo: delete this once we figure out how run Oracle try-catch as well.
```
Comment (fil512) [inline hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTask.java:165]: is this a todone?

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTask.java [hunk 3]
```diff
@@ -161,24 +162,62 @@ String generateSql() {
...
 					if (myMetadataSource.isOnlineIndexSupported(getConnectionProperties())) {
-						msSqlOracleOnlineClause = " ONLINE DEFERRED INVALIDATION";
-					}
-					break;
-				case MSSQL_2012:
```
Comment (fil512) [inline hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTask.java:168]: consider adding:
```
\\ case MSSQL_2012: 
\\ this case is handled below
```

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTask.java [hunk 4]
```diff
@@ -161,24 +162,62 @@ String generateSql() {
...
+		if (myOnline && DriverTypeEnum.MSSQL_2012 == getDriverType()) {
+			sql = buildOnlineCreateWithTryCatchFallback(bareCreateSql);
+		} else {
+			sql = bareCreateSql;
+		}
 		return sql;
 	}
 
+	/**
+	 * Wrap a Sql Server create index in a try/catch to try it first ONLINE
```
Comment (fil512) [inline hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTask.java:187]: one disadvantage of doing it this way is there is no warning to the user that their system will be frozen for an indeterminate amount of time

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTask.java [hunk 5]
```diff
@@ -161,24 +162,62 @@ String generateSql() {
...
+	 * -- so the CATCH can catch it.
+	 * }
+	 *
+	 * @param bareCreateSql
+	 * @return
+	 */
+	private static @Nonnull String buildOnlineCreateWithTryCatchFallback(String bareCreateSql) {
+		// Some "Editions" of Sql Server do not support ONLINE.
+		// @format:off
+		return "BEGIN TRY -- try first online, without locking the table \n"
```
Comment (fil512) [inline hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTask.java:211]: this is very cool. i did not know this was a thing.

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/MetadataSource.java [hunk 6]
```diff
@@ -32,16 +32,23 @@ public class MetadataSource {
...
 
+		// todo: delete this once we figure out how run Oracle try-catch as well.
 		switch (theConnectionProperties.getDriverType()) {
 			case POSTGRES_9_4:
 			case COCKROACHDB_21_1:
 				return true;
 			case MSSQL_2012:
+				// use a deny-list instead of allow list, so we have a better failure mode for new/unknown versions.
+				// Better to fail in dev than run with a table lock in production.
 				String mssqlEdition = getEdition(theConnectionProperties);
```
Comment (fil512) [inline hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/MetadataSource.java:43]: nitpick: code might read clearer if getEdition returned an Optional. Giving null special meaning can be confusing.

File: hapi-fhir-sql-migrate/src/test/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTaskITTestSuite.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+
+/**
+ * Integration tests for AddIndexTask.
+ */
+public interface AddIndexTaskITTestSuite extends BaseMigrationTaskTestSuite {
+
+	@Test
+	default void testAddIndexOnline_createsIndex() throws SQLException {
+		// given
+		Builder builder = getSupport().getBuilder();
```
Comment (fil512) [inline hapi-fhir-sql-migrate/src/test/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTaskITTestSuite.java:21]: did this test fail before you added this test?

File: hapi-fhir-sql-migrate/src/test/java/ca/uhn/fhir/jpa/migrate/taskdef/containertests/AbstractMigrationTaskSuite.java [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import org.junit.jupiter.api.Test;
+
+import jakarta.annotation.Nonnull;
+
+import static ca.uhn.fhir.jpa.migrate.taskdef.containertests.BaseMigrationTaskTestSuite.Support;
+
+/**
+ * Collects all our task suites in a single class so we can run them on each engine.
+ */
+public abstract class AbstractMigrationTaskSuite {
```
Comment (fil512) [inline hapi-fhir-sql-migrate/src/test/java/ca/uhn/fhir/jpa/migrate/taskdef/containertests/AbstractMigrationTaskSuite.java:17]: BaseMigrationTaskSuite?

File: hapi-fhir-sql-migrate/src/test/java/ca/uhn/fhir/jpa/migrate/taskdef/containertests/SqlServerAzureMigrationTests.java [hunk 9]
```diff
@@ -0,0 +1,26 @@
...
+
+public class SqlServerAzureMigrationTests extends AbstractMigrationTaskSuite {
+	@RegisterExtension
+	static TestContainerDatabaseMigrationExtension ourContainerExtension =
+		new TestContainerDatabaseMigrationExtension(
+			DriverTypeEnum.MSSQL_2012,
+			new MSSQLServerContainer<>(
+				DockerImageName.parse("mcr.microsoft.com/azure-sql-edge:latest")
+					.asCompatibleSubstituteFor("mcr.microsoft.com/mssql/server"))
+				.withEnv("ACCEPT_EULA", "Y")
```
Comment (fil512) [inline hapi-fhir-sql-migrate/src/test/java/ca/uhn/fhir/jpa/migrate/taskdef/containertests/SqlServerAzureMigrationTests.java:18]: suggestion: comment to explain what this PID is about

Comment (fil512) [reply hapi-fhir-sql-migrate/src/test/java/ca/uhn/fhir/jpa/migrate/taskdef/containertests/SqlServerAzureMigrationTests.java:18]: question: I don't see anywhere that we assert the sql try/catch worked. should we consider sniffing logs to validate it performed as expected?

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTask.java (hunk 2 above)
Comment (michaelabuckley) [reply hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTask.java:165]: No - I don't have an easy way to test Oracle, so I left the Oracle code behind.

File: hapi-fhir-sql-migrate/src/test/java/ca/uhn/fhir/jpa/migrate/taskdef/containertests/SqlServerAzureMigrationTests.java (hunk 9 above)
Comment (michaelabuckley) [reply hapi-fhir-sql-migrate/src/test/java/ca/uhn/fhir/jpa/migrate/taskdef/containertests/SqlServerAzureMigrationTests.java:18]: I tested that manually.  See the javadoc for `AddIndexTask.buildOnlineCreateWithTryCatchFallback()`.   There's no convenient way to get logs from inside the sql.

File: hapi-fhir-sql-migrate/src/test/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTaskITTestSuite.java (hunk 7 above)
Comment (michaelabuckley) [reply hapi-fhir-sql-migrate/src/test/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTaskITTestSuite.java:21]: No.  As you say, this approach fails over silently to a table lock.  The good news is, anyone using Sql Server Standard has this as a global limitation.  I think a note in our upgrade docs can address this.

File: hapi-fhir-sql-migrate/src/test/java/ca/uhn/fhir/jpa/migrate/taskdef/containertests/AbstractMigrationTaskSuite.java (hunk 8 above)
Comment (michaelabuckley) [reply hapi-fhir-sql-migrate/src/test/java/ca/uhn/fhir/jpa/migrate/taskdef/containertests/AbstractMigrationTaskSuite.java:17]: BaseCollectedMigrationTaskSuite

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/MetadataSource.java (hunk 6 above)
Comment (michaelabuckley) [reply hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/MetadataSource.java:43]: I plan to delete this class.

File: hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTask.java (hunk 3 above)
Comment (michaelabuckley) [reply hapi-fhir-sql-migrate/src/main/java/ca/uhn/fhir/jpa/migrate/taskdef/AddIndexTask.java:168]: done

**Review by tadgh — APPROVED** [review]

**Pull request merged**